### PR TITLE
[TRTLLM-16316][feat] Generalize Runtime Sleep/Wakeup for LLM and `trtllm-serve`

### PR DESCRIPTION
**Pull request by Funatiq**
## Dev Engineer Review

- Adds synchronous and asynchronous runtime-memory APIs: `get_memory_status()`, `release()`, and `resume()`.
- Adds authenticated runtime-control endpoints and `trtllm-serve --enable_sleep_mode`.
- Adds tag validation, transition handling, V1 prefix-cache invalidation, and KVCacheManagerV2 rejection.
- Main risks are API compatibility, authentication, replay protection, transition failures, and consistency across runtime paths.
- CI remains unresolved: four L0 pipelines failed, two were unstable, and the latest run has no result.

## QA Engineer Review

- Adds coverage for runtime-control authentication, replay protection, HTTP errors, runtime-memory APIs, sleep-mode CLI validation, transition failures, prefix-cache invalidation, KVCacheManagerV2 rejection, and collective RPC behavior.
- Updates CPU and H100 integration test lists for the new runtime-control, runtime-memory, serve sleep-mode, and Ray sleep-state tests.
- Coverage verdict: **needs follow-up** because CI failures remain unresolved and the latest run has no result.

### Per-File QA Perspective

- `docs/source/features/sleep-wakeup.md`: Verify restore-mode documentation.
- `docs/source/index.rst`: Verify the new guide is listed.
- `tensorrt_llm/_torch/async_llm.py`: Verify async status, release, resume, defaults, and tag normalization.
- `tensorrt_llm/_torch/pyexecutor/executor_request_queue.py`: Verify atomic status snapshots.
- `tensorrt_llm/_torch/pyexecutor/py_executor.py`: Verify status reporting, prefix-cache invalidation, and V2 rejection.
- `tensorrt_llm/commands/serve.py`: Verify sleep-mode restrictions and API-key requirements.
- `tensorrt_llm/executor/base_worker.py`: Verify state filtering, validation, and no-op behavior.
- `tensorrt_llm/executor/proxy.py`: Verify multi-rank status forwarding.
- `tensorrt_llm/executor/ray/gpu_worker.py`: Verify transition success and failure handling.
- `tensorrt_llm/llmapi/__init__.py`: Verify public exports.
- `tensorrt_llm/llmapi/llm.py`: Verify synchronous runtime-memory operations and validation.
- `tensorrt_llm/llmapi/llm_args.py`: Verify status validation, defaults, and tag ordering.
- `tensorrt_llm/serve/openai_protocol.py`: Verify optional tag parsing.
- `tensorrt_llm/serve/openai_server.py`: Verify authentication, routes, dispatch, and error mapping.
- `tensorrt_llm/serve/runtime_control_auth.py`: Verify signing, timestamp checks, and nonce replay protection.
- `tests/integration/test_lists/test-db/l0_cpu.yml`: Adds CPU CI entries for runtime-control, runtime-memory, and sleep-mode tests.
- `tests/integration/test_lists/test-db/l0_h100.yml`: Adds the Ray sleep-state test to H100 CI.
- `tests/unittest/_torch/executor/test_executor_request_queue.py`: Covers partial memory restoration.
- `tests/unittest/_torch/multi_gpu/test_mpi_sleep_wakeup.py`: Covers public release/resume and partial restoration.
- `tests/unittest/api_stability/references/llm.yaml`: Covers API stability references for runtime-memory methods.
- `tests/unittest/api_stability/references/trtllm_serve_cli.yaml`: Covers the new CLI flag reference.
- `tests/unittest/executor/test_ray_sleep_state.py`: Covers Ray transition success and failure paths.
- `tests/unittest/executor/test_sleep_collective_rpc_guards.py`: Covers V2 rejection, V1 invalidation, retries, and RPC forwarding.
- `tests/unittest/llmapi/apps/test_anthropic_routes.py`: Updates route-test setup for runtime-control state.
- `tests/unittest/llmapi/test_runtime_control.py`: Covers authentication, replay protection, concurrency, configuration conflicts, and HTTP errors.
- `tests/unittest/llmapi/test_runtime_memory_api.py`: Covers synchronous and asynchronous runtime-memory APIs.
- `tests/unittest/llmapi/test_serve_sleep_mode.py`: Covers sleep-mode CLI validation and configuration.
- `tests/unittest/llmapi/test_async_llm.py`: Excludes initialization-only memory types from async release and resume tests.

## Description

- Add documentation for Sleep and Wakeup feature.
- Generalize API from AsyncLLM to sync LLM and trtllm-serve.
- Fix KV cache manager V1: invalidate prefix cache before destructive sleep
- Reject KV cache sleep with manager V2 for now. Clean fix is a follow-up.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/llmapi/llm.py [hunk 1]
```diff
@@ -1940,6 +1946,109 @@ def _collective_rpc(
...
+                Memory types to release. When omitted, release every
+                sleep-managed runtime memory type.
+
+        Returns:
+            None: The call returns after all selected memory has been released.
+        """
+        self._check_runtime_memory_enabled()
+        normalized = self._normalize_runtime_memory_tags(
+            tags, default_tags=_RUNTIME_MEMORY_TYPES)
+        self._collective_rpc("sleep", ([tag.value for tag in normalized], ))
```
Comment (nv-xtf) [inline tensorrt_llm/llmapi/llm.py:2027]: Could you clarify how KV-cache release/resume works across V1 and V2? With the default NONE mode, V1 appears to discard KV contents without invalidating prefix-reuse metadata, so a later reuse hit could read invalid KV. V2’s main pools use a separate allocator and appear to bypass tag-based release, potentially remaining resident while kv_cache is reported as parked. Is there an invalidation or release path I’m missing? The MPI sleep tests disable block reuse and don’t explicitly cover V2.

**Review by nv-xtf — CHANGES_REQUESTED** [review]

**Review by weiminwang-nv — APPROVED** [review]

File: tensorrt_llm/llmapi/llm.py (hunk 1 above)
Comment (Funatiq) [reply tensorrt_llm/llmapi/llm.py:2055]: Yes, you are right. I was assuming this is already implemented correctly. I added V1 reuse invalidation and an assert to disallow V2 for now. I suggest a clean fix for V2 as a followup.

File: tests/unittest/_torch/multi_gpu/test_mpi_sleep_wakeup.py [hunk 2]
```diff
@@ -123,7 +123,10 @@ def test_mpi_sleep_wakeup_tp2(process_gpu_memory_info_available):
                 "Check pynvml setup or GPU visibility."
             )
 
-        llm._collective_rpc("sleep", (sleep_tags,))
+        llm.release(sleep_tags)
+        status = llm.get_memory_status()
+        assert status.state == "parked"
+        assert status.parked_tags == sleep_tags
```
[the remark this answers is not in the document]
Comment (Funatiq) [reply tests/unittest/_torch/multi_gpu/test_mpi_sleep_wakeup.py:129]: Addressed in commit ac56287. Tags are sorted lexicographically now.

File: docs/source/features/sleep-wakeup.md [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+Release can preserve allocation contents in host memory and free their GPU
+physical memory. Resume allocates GPU physical memory again, maps it at the
+same virtual addresses, and restores the saved contents. The virtual addresses
+remain reserved while the runtime is parked. This is an in-process memory
+checkpoint: it is not written to disk and does not survive process termination
+or support restart in a different process.
+
+```{warning}
+Sleep and wakeup are prototype APIs and are subject to change. They are
+available only with the PyTorch backend. AutoDeploy is not supported.
```
Comment (arysef) [inline docs/source/features/sleep-wakeup.md:21]: nit: since AD is deprecated, we don't need to explicitly mention it

**Review by arysef — APPROVED** [review]

File: tensorrt_llm/executor/ray/gpu_worker.py [hunk 4]
```diff
@@ -315,17 +315,34 @@ def sleep(self, sleep_tags: List[str]):
...
                 "Sleep feature is not enabled, please set sleep_config in the LLM arguments."
             )
+        tags = [ExecutorMemoryType(tag) for tag in sleep_tags]
+        tags = self._prepare_sleep_tags(tags)
+        if not tags:
+            return
+        self._validate_sleep_tags(tags)
+
+        logger.info(f"Sleep: {tags}")
+        self.engine.begin_sleep_transition(tags)
```
Comment (chienchunhung) [inline tensorrt_llm/executor/ray/gpu_worker.py]: IIUC We might need to move `begin_sleep_transition()` even before entering `control_action()`. The decorator queues the control sentinel and drains work before this transition executes. Independent submission threads can still admit requests behind that sentinel, which then enter the scheduler after memory is released. This leaves the newly documented admission guarantee incomplete.

File: tensorrt_llm/serve/openai_server.py [hunk 5]
```diff
@@ -752,6 +774,9 @@ def __init__(
         self._internal_disagg_auth_key = internal_disagg_auth_key
         self._enable_rl_control_endpoints = enable_rl_control_endpoints
         self._rl_control_api_key = rl_control_api_key
+        self._enable_runtime_control_endpoints = enable_runtime_control_endpoints
+        self._runtime_control_api_key = runtime_control_api_key
+        self._runtime_control_replay_cache = RuntimeControlReplayCache()
```
Comment (chienchunhung) [inline tensorrt_llm/serve/openai_server.py]: Each frontend may have an independent replay cache. A signed release accepted by frontend `A` can therefore be replayed through frontend `B` after a legitimate resume, parking the executor again. We should probably coordinate across frontends for the same executor, or reject sleep mode with `num_serve_frontends > 1`.

File: tensorrt_llm/llmapi/llm.py [hunk 6]
```diff
@@ -1968,6 +1974,109 @@ def _collective_rpc(
...
```
Comment (chienchunhung) [inline tensorrt_llm/llmapi/llm.py]: Please support runtime-memory dispatch through `GenerationExecutorWorker`. TP1 with `gather_generation_logits=True` selects this worker under the default orchestrator, but its inheritance chain has no `collective_rpc`. Release, status, and resume would therefore raise “does not support collective RPC.”

File: tensorrt_llm/llmapi/llm.py [hunk 7]
```diff
@@ -1968,6 +1974,109 @@ def _collective_rpc(
...
+        Args:
+            tags (Sequence[tensorrt_llm.llmapi.ExecutorMemoryType | str], optional):
+                Memory types to restore. When omitted, restore every currently
+                parked memory type.
+
+        Returns:
+            None: The call returns after all selected memory has been restored.
+        """
+        self._check_runtime_memory_enabled()
+        default_tags = self.get_memory_status().parked_tags
```
Comment (chienchunhung) [inline tensorrt_llm/llmapi/llm.py]: A failed release can leave failed with no published parked tags; default resume skips the worker guard and returns success while admission remains closed. We need to validate the status state before treating empty `parked_tags` as a successful no-op.

**Review by chienchunhung — COMMENTED** [review]
Thanks for the PR! I left a few inline comments.
